Services/Statistics/SimulationStatisticsAgent.cs

```diff
@@ -0,0 +1,86 @@
+{
+    public interface ISimulationStatistics
+    {
+        Task<SimulationStatisticsModel> GetSimulationStatisticsAsync(string simulationId);
+        Task CreateOrUpdateAsync(string simulationId, SimulationStatisticsModel statistics);
+    }
+
+    public class SimulationStatisticsAgent : ISimulationStatistics
+    {
+        private readonly IStorageRecords simulationStatisticsStorage;
+        private readonly IClusterNodes clusterNodes;
+        private readonly ISimulations simulations;
+        private readonly IServicesConfig config;
+        private readonly ILogger log;
+
+        public SimulationStatisticsAgent(IServicesConfig config,
+            IClusterNodes clusterNodes,
+            ISimulations simulations,
+            IFactory factory,
+            ILogger logger)
+        {
+            this.clusterNodes = clusterNodes;
+            this.simulations = simulations;
+            this.config = config;
+            this.log = logger;
+            this.simulationStatisticsStorage = factory.Resolve<IStorageRecords>().Init(config.StatisticsStorage);
+        }
+
+        public async Task<SimulationStatisticsModel> GetSimulationStatisticsAsync(string simulationId)
+        {
+            string sqlCondition = " CONTAINS(ROOT.id, @simulationId)";
+            SqlParameter[] sqlParameters = new[] { new SqlParameter { Name = "@simulationId", Value = simulationId } };
+
+            var simRecords = (await this.simulationStatisticsStorage.GetAsync(sqlCondition, sqlParameters))
+                .Select(p => JsonConvert.DeserializeObject<SimulationStatisticsRecord>(p.Data))
+                .ToList();
+
+            SimulationStatisticsModel statistics = new SimulationStatisticsModel();
+            statistics.TotalMessagesSent = simRecords.Sum(a => a.Statistics.TotalMessagesSent);
+            statistics.FailedDeviceConnectionsCount = simRecords.Sum(a => a.Statistics.FailedDeviceConnectionsCount);
+            statistics.FailedDeviceTwinUpdatesCount = simRecords.Sum(a => a.Statistics.FailedDeviceTwinUpdatesCount);
+            statistics.FailedMessagesCount = simRecords.Sum(a => a.Statistics.FailedMessagesCount);
+
+            return statistics;
+        }
+        
+        public async Task CreateOrUpdateAsync(string simulationId, SimulationStatisticsModel statistics)
+        {
+            var nodeId = this.clusterNodes.GetCurrentNodeId();
+            var statisticsRecordId = this.GetStatisticsRecordId(simulationId, nodeId);
+
+            var stats = new SimulationStatisticsRecord
+            {
+                NodeId = nodeId,
+                SimulationId = simulationId,
+                Statistics = statistics
+            };
+
+            var statsRecord = new StorageRecord
+            {
+                Id = statisticsRecordId,
+                Data = JsonConvert.SerializeObject(stats)
+            };
+
+            await this.simulationStatisticsStorage.CreateAsync(statsRecord);
```

Comment:
You'll probably want some try/catch logic here, too, no?

Reply:
thats right, added